prow/metrics/prowjobs/collector.go

```diff
@@ -0,0 +1,172 @@
+		default:
+			return
+		}
+
+		switch newJob.Status.State {
+
+		case prowapi.FailureState, prowapi.SuccessState, prowapi.ErrorState, prowapi.AbortedState:
+			if newJob.Status.CompletionTime == nil {
+				return
+			}
+			newTime = *newJob.Status.CompletionTime
+		case prowapi.PendingState:
+			if newJob.Status.PendingTime == nil {
+				return
+			}
+			newTime = *newJob.Status.PendingTime
+		default:
+			return
+		}
+
+		histogram.Observe(newTime.Sub(oldTime.Time).Seconds())
+	}
+}
+
+// NewProwJobLifecycleCollector creates histograms which can track the timespan between ProwJob state transitions.
+// The histograms are based on the job name, the old job state and the new job state.
+// Data is collected by hooking itself into the prowjob informer.
+// The collector will never record the same state transition twice, even if reboots happen.
+func NewProwJobLifecycleCollector(informer cache.SharedIndexInformer) *prowJobLifecycleCollector {
+	lifecycleCollector := &prowJobLifecycleCollector{
+		HistogramVec: newHistogramVec(),
+	}
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(oldJob, newJob interface{}) {
+			lifecycleCollector.Update(oldJob.(*prowapi.ProwJob), newJob.(*prowapi.ProwJob))
+		},
+	})
+	return lifecycleCollector
+}
+
+func getJobLabel(oldJob *prowapi.ProwJob, newJob *prowapi.ProwJob) jobLabel {
+	jl := jobLabel{jobNamespace: newJob.Namespace, jobName: newJob.Spec.Job, jobType: string(newJob.Spec.Type), state: string(newJob.Status.State), last_state: string(oldJob.Status.State)}
+
+	if newJob.Spec.Refs != nil {
+		jl.org = newJob.Spec.Refs.Org
+		jl.repo = newJob.Spec.Refs.Repo
+		jl.baseRef = newJob.Spec.Refs.BaseRef
+	} else if len(newJob.Spec.ExtraRefs) > 0 {
+		jl.org = newJob.Spec.ExtraRefs[0].Org
+		jl.repo = newJob.Spec.ExtraRefs[0].Repo
+		jl.baseRef = newJob.Spec.ExtraRefs[0].BaseRef
+	}
+
+	return jl
+}
+
+type jobLabel struct {
+	jobNamespace string
+	jobName      string
+	jobType      string
+	last_state   string
+	state        string
+	org          string
+	repo         string
+	baseRef      string
+}
+
+func (jl *jobLabel) values() []string {
+	return []string{jl.jobNamespace, jl.jobName, jl.jobType, jl.last_state, jl.state, jl.org, jl.repo, jl.baseRef}
+}
+
+func newHistogramVec() *prometheus.HistogramVec {
+	return prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "prow_job_runtime_seconds",
+			Buckets: []float64{
+				time.Minute.Seconds() / 2,
+				(1 * time.Minute).Seconds(),
+				(2 * time.Minute).Seconds(),
+				(5 * time.Minute).Seconds(),
+				(10 * time.Minute).Seconds(),
+				(1 * time.Hour).Seconds() / 2,
+				(1 * time.Hour).Seconds(),
+				(2 * time.Hour).Seconds(),
+				(3 * time.Hour).Seconds(),
+				(4 * time.Hour).Seconds(),
+				(5 * time.Hour).Seconds(),
+				(6 * time.Hour).Seconds(),
+				(7 * time.Hour).Seconds(),
+				(8 * time.Hour).Seconds(),
+				(9 * time.Hour).Seconds(),
+				(10 * time.Hour).Seconds(),
+			},
+		},
+		[]string{
```

Comment:
So I still don't really understand the worry about "the number of histograms". Based on the labels we have listed here, we have bounded cardinality. You don't create a histogram per job execution (one run of a job) but for all executions of a job across time.

Reply:
Yes, my only worry would be the size of the document which prometheus scrapes. Since the exporter does not stop to export timers ever, it could become a very expensive scrape targets.